client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -20,13 +21,42 @@ private final class PoolManager[F[_], A <: Connection[F]](
   private[this] val logger = getLogger(classOf[PoolManager[F, A]])
 
   private var isClosed = false
-  private var allocated = 0
-  private val idleQueue = new mutable.Queue[A]
-  private val waitQueue = new mutable.Queue[Waiting]
+  private var curTotal = 0
+  private var allocated = mutable.Map.empty[RequestKey, Int]
+  private val idleQueue = mutable.Map.empty[RequestKey, mutable.Queue[A]]
+  private val waitQueue = mutable.Queue.empty[Waiting]
 
   private def stats =
     s"allocated=$allocated idleQueue.size=${idleQueue.size} waitQueue.size=${waitQueue.size}"
 
+  @inline
+  private def getMaxConnections(key: RequestKey): Int =
+    maxConnectionsPerRequestKey.getOrElse(key, maxTotal)
+
+  @inline
+  private def getConnectionFromQueue(key: RequestKey): Option[A] =
+    try {
+      Some(idleQueue(key).dequeue())
```

Comment:
I'd probably rather do an empty check than catch exceptions.

Reply:
Yeah, I think it's a silly optimisation (if it even was one).